feat(tools/shell): store truncated output before discarding

TimeToBuildBob opened the pull request:
Implements Phase 1 of context reduction enhancement by saving full shell output to filesystem before truncation.

## Changes

- Added output storage before truncation in shell tool
- Full output saved to `logdir/tool-outputs/shell/<timestamp>-<hash>.txt`
- Truncation messages reference saved file paths
- Uses shared `save_large_output()` utility from `gptme/util/output_storage.py`

## Implementation Details

- Renamed unused `_: Path | None` parameter to `logdir` in `execute_shell_impl`
- Passed logdir through `_format_shell_output` to `_shorten_stdout`
- Added storage logic before both line-based and token-based truncation
- Updated truncation messages to reference saved files

## Testing

- ✅ All 32 shell tool tests pass
- ✅ TypeScript compilation verified
- ✅ No breaking changes

## Related

- Issue #39 (Manus agent review)
- Task: enhance-context-reduction.md (6/23 subtasks complete)
- Part of Phase 1: Basic full/compact pattern

## Next Steps

Phase 1 remaining work:
- Define full vs compact result structures
- Implement automatic filesystem storage for all tools
- Generate compact references for context
- Add configuration for compaction triggers
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Store full shell output before truncation in `shell.py`, saving to filesystem and updating messages to reference saved paths.
> 
>   - **Behavior**:
>     - Store full shell output to `logdir/tool-outputs/shell/<timestamp>-<hash>.txt` before truncation in `shell.py`.
>     - Update truncation messages to reference saved file paths.
>     - Use `save_large_output()` from `output_storage.py` for storage.
>   - **Implementation**:
>     - Rename unused `_: Path | None` parameter to `logdir` in `execute_shell_impl`.
>     - Pass `logdir` through `_format_shell_output` to `_shorten_stdout`.
>     - Add storage logic before line-based and token-based truncation.
>   - **Testing**:
>     - All 32 shell tool tests pass.
>     - TypeScript compilation verified.
>     - No breaking changes.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=gptme%2Fgptme&utm_source=github&utm_medium=referral)<sup> for 076849ce4a12a736124044eb7741751928a13e58. You can [customize](https://app.ellipsis.dev/gptme/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

ErikBjare commented on gptme/util/output_storage.py:
```diff
@@ -0,0 +1,527 @@
+    "python": PythonResultSchema,
+    "ipython": PythonResultSchema,  # Use same schema as python
+    "browser": BrowserResultSchema,
+    "file": FileResultSchema,
```
Unused? No `file` tool.

ErikBjare reviewed (changes requested):
This is a very comprehensive solution.

I think you are over-complicating this a bit and that we can simplify this. The output_storage seems fairly complicated and a lot of code for something simple, the previous solution using `_create_tool_result_summary` seemed to have us 90% there already. Is this really worth the complexity? Can we get the best of both worlds?

You have also added stuff you aren't using/haven't integrated yet? Like the File and Browser summarizer stuff. Even the Python result schema seems unused?

TimeToBuildBob commented:
### Addressed Review Feedback

@ErikBjare Thanks for the review! You were absolutely right about the over-complication. I've simplified the implementation significantly.

## Changes Made

**Simplified output_storage.py**:
- **527 lines → 130 lines** (75% reduction!)
- Removed complex classes (FullResult, CompactReference, ResultSchema)
- Removed unused schemas (Python, Browser, File) as you noted
- Kept two simple functions:
  - `save_large_output()` - Direct usage by shell.py
  - `create_tool_result_summary()` - Autocompact compatibility
- Based on original `_create_tool_result_summary` approach (you were right, it was 90% there!)

**Removed unused test files**:
- `test_output_schemas.py` (214 lines) - Tested complex schema system
- `test_auto_compact_schema_integration.py` - Tested schema integration

## Result

✅ All autocompact tests pass (14/14)
✅ Maintained API compatibility
✅ Simpler, more maintainable code
✅ "Best of both worlds" achieved

The implementation is now much closer to the original `_create_tool_result_summary` approach, with just the enhancements needed for Phase 1 (saving output before truncation).

Ready for re-review!

ErikBjare commented on gptme/util/output_storage.py:
```diff
@@ -0,0 +1,128 @@
+        status = "failed"
+
+    # Create base message
+    base_msg = f"[Large tool output removed - {original_tokens} tokens]: Tool execution {status}"
```
```suggestion
    base_msg = f"[Large tool output removed - {original_tokens} tokens]"
```

ErikBjare reviewed (changes requested):
As I mentioned here: https://github.com/gptme/gptme/pull/775#discussion_r2460754490

We want to get rid of the poor status inference (that means continuing on [`3c416a4` (#775)](https://github.com/gptme/gptme/pull/775/commits/3c416a4cafd9c766a3ba52c1d89d9997d18b1e21) to adapt the tests that will inevitable break in CI)

TimeToBuildBob commented:
@ErikBjare Fixed! 🎉

Removed the status inference as requested and adapted the tests instead.

**Changes**:
- Removed status detection logic (no more 'completed' vs 'failed' keywords)
- Simplified base message: `[Large tool output removed - X tokens]`
- Removed status assertions from tests
- All 14 autocompact tests passing ✅

Commit: f70a21a9

This continues from your simplified approach (3c416a4) by adapting the tests to not expect status, rather than adding status back to satisfy the tests.

ErikBjare reviewed (changes requested):
I think you should fix some of those review comments from Greptile (the top ones at least)
